handler_storage.js

```diff
@@ -1,101 +1,99 @@
+    this.constraints = []
+    this.constrainedHandlers = [] // unoptimized list of handler objects for which the fast matcher function will be compiled
     this.constrainedHandlerStores = null
   }
 
-  getHandler (constraints) {
-    return this.handlers.filter(handler => deepEqual(constraints, handler.constraints))[0]
+  hasHandler (constraints) {
+    if (Object.keys(constraints).length === 0) {
+      return this.unconstrainedHandler !== null
+    }
+
+    for (const handler of this.constrainedHandlers) {
+      if (deepEqual(constraints, handler.constraints)) {
+        return true
+      }
+    }
+    return false
   }
 
   // This is the hot path for node handler finding -- change with care!
-  getMatchingHandler (constrainer, derivedConstraints) {
+  getMatchingHandler (derivedConstraints) {
     if (derivedConstraints === undefined) {
       return this.unconstrainedHandler
     }
 
-    if (this.hasConstraints) {
-      // This node is constrained, use the performant precompiled constraint matcher
-      return this._getHandlerMatchingConstraints(constrainer, derivedConstraints)
-    }
+    const handler = this._getHandlerMatchingConstraints(derivedConstraints)
 
-    // This node doesn't have any handlers that are constrained, so it's handlers probably match. Some requests have constraint values that *must* match however, like version, so check for those before returning it.
-    if (!derivedConstraints.__hasMustMatchValues) {
+    if (handler === null && !derivedConstraints.__hasMustMatchValues) {
       return this.unconstrainedHandler
     }
 
-    return null
+    return handler
   }
 
-  addHandler (handler, params, store, constraints) {
-    if (!handler) return
-    assert(!this.getHandler(constraints), `There is already a handler with constraints '${JSON.stringify(constraints)}' and method '${this.method}'`)
-
+  addHandler (handler, params, store, constrainer, constraints) {
     const handlerObject = { handler, params, constraints, store: store || null, paramsLength: params.length }
 
-    this.handlers.push(handlerObject)
-    // Sort the most constrained handlers to the front of the list of handlers so they are tested first.
-    this.handlers.sort((a, b) => Object.keys(a.constraints).length - Object.keys(b.constraints).length)
-
-    if (Object.keys(constraints).length > 0) {
-      this.hasConstraints = true
-    } else {
+    if (Object.keys(constraints).length === 0) {
       this.unconstrainedHandler = handlerObject
+      return
+    }
+
+    for (const constraint of Object.keys(constraints)) {
+      if (!this.constraints.includes(constraint)) {
+        if (constraint === 'version') {
+          // always check the version constraint first as it is the most selective
+          this.constraints.unshift(constraint)
+        } else {
+          this.constraints.push(constraint)
+        }
+      }
     }
 
-    if (this.hasConstraints && this.handlers.length > 32) {
+    if (this.constrainedHandlers.length >= 32) {
```

Comment:
why this has became a `>=` check instead of a `>` check?

Reply:
Because before we made this check after inserting a handler into the handler's array and now we check it before inserting it.